platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,1026 @@
+   to receive the updated state containing the TSS key material.
+
+#### Startup for Existing Networks
+
+##### The Release Containing the `tssEncryptionKey`
+
+This release must occur with or after the complete deployment of the TSS-Roster proposal.
+
+This release contains all the state data structures needed to store the public `tssEncryptionKey` in the state and all
+code needed to support storing and retrieving the private `tssEncryptionKey` on disk and managing the `tssEncryptionKey`
+life cycle, including key rotation.
+
+The release supporting the addition of the `tssEncryptionKey` to the roster also provides modified HAPI
+transactions for updating the address book with a node's `tssEncryptionKey`. Until the next software upgrade, only
+the candidate roster will have the `tssEncryptionKey` possibly set in roster entries.
+
+The only change to the startup sequence is that the `tssEncryptionKey` private key is read from disk, provided to
+the platform through the `PlatformBuilder` and validated against the public `tssEncryptionKey` in the active roster.
+A mismatch is a critical error that will cause the node to shut down.
+
+##### The Release Containing the TSS Bootstrap Process
+
+This release must be at least 1 release after the `tssEncryptionKey` release.
+
+This sequence will work for this release and all subsequent releases, unless the
+startup sequence is modified by subsequent releases.<br/>
+The following startup sequence will occur:
+
+1. The app will read the state and cryptography from disk (including support for
+   reading the private `tssEncryptionKey`)
+2. The app will validate the cryptography and state.
+3. If a software upgrade is detected,
+   1. If the existing active roster and candidate roster do not have complete
+      key material and the network does not yet have a ledger id, the node will
+      adopt the candidate roster immediately on software upgrade.
+      - This will retain the pre-ledger-id behavior of the network until the
+        ledger id can be created for the first time.
+   2. If the candidate roster exists, has key material, and the number of yes
+      votes passes the threshold for adoption, then the candidate roster
+      will be adopted and become the new active roster.
+      1. Validation Check: if the state already has a ledger id set, the ledger
+         id recovered from the key material must match the ledger id
+         in the state.
+      2. If there is no ledger id in the state and the previously active roster
+         has no key material, the ledger id recovered from the candidate roster
+         key material will be set in the state as the new ledger id.
+   3. In all other cases the existing active roster will remain the
+      active roster.
+4. The application will provide the platform with the active roster to use.
+5. The platform will continue through its normal startup sequence.
```

Comment:
Is only during upgrade, a candidate roster can become an active roster?

Reply:
At the moment, yes. Once we have dynamic address book phase 3, the update will be triggered by the first transaction after midnight.